src/block_cache/block_io.c

```diff
@@ -355,11 +378,41 @@ __wt_bt_write(WT_SESSION_IMPL *session, WT_ITEM *buf, uint8_t *addr, size_t *add
         WT_STAT_SESSION_INCRV(session, write_time, time_diff);
     }
 
+    /*
+     * The page image must have a proper write generation number before writing it to disk. The page
+     * images that are created during recovery may have the write generation number less than the
+     * btree base write generation number, so don't verify it.
+     */
+    dsk = ip->mem;
+    WT_ASSERT(session, dsk->write_gen != 0);
+
     WT_STAT_CONN_DATA_INCR(session, cache_write);
     WT_STAT_CONN_DATA_INCRV(session, cache_bytes_write, dsk->mem_size);
     WT_STAT_SESSION_INCRV(session, bytes_write, dsk->mem_size);
     (void)__wt_atomic_add64(&S2C(session)->cache->bytes_written, dsk->mem_size);
 
+    /*
+     * Store a copy of the compressed buffer in the block cache.
+     *
+     * Optional if the write is part of a checkpoint. Hot blocks get written and over-written a lot
+     * as part of checkpoint, so we don't want to cache them, because (a) they are in the in-memory
+     * cache anyway, and (b) they are likely to be overwritten again in the next checkpoint. Writes
+     * that are not part of checkpoint I/O are done in the service of eviction. Those are the blocks
+     * that the in-memory cache would like to keep but can't, and we definitely want to keep them.
+     *
+     * Optional on normal writes (vs. reads) if the no-write-allocate setting is on.
+     *
+     * Ignore the final checkpoint writes.
+     */
```

Comment:
What are those final checkpoint writes and why do we ignore them?

Reply:
The final checkpoint write is the tree's root page. It's read at file open, and never read through the cache, so there's no point in writing it into the cache.